httomo/methods_database/packages/external/httomolibgpu/supporting_funcs/recon/algorithm.py

```diff
@@ -66,28 +66,54 @@ def _calc_memory_bytes_FBP(
     dtype: np.dtype,
     **kwargs,
 ) -> Tuple[int, int]:
-    DetectorsLengthH = non_slice_dims_shape[1]
-    # calculate the output shape
+    det_width = non_slice_dims_shape[1]
     output_dims = _calc_output_dim_FBP(non_slice_dims_shape, **kwargs)
 
-    in_slice_size = np.prod(non_slice_dims_shape) * dtype.itemsize
-    filter_size = (DetectorsLengthH // 2 + 1) * np.float32().itemsize
-    freq_slice = (
-        non_slice_dims_shape[0] * (DetectorsLengthH // 2 + 1) * np.complex64().itemsize
+    input_slice_size = np.prod(non_slice_dims_shape) * dtype.itemsize
+    filtered_input_slice_size = np.prod(non_slice_dims_shape) * np.float32().itemsize
+    # astra backprojection will generate an output array
+    recon_output_size = np.prod(output_dims) * np.float32().itemsize
+
+    # filter needs to be created once and substracted from the total memory
+    filter_size = (det_width // 2 + 1) * np.float32().itemsize
+
+    # this stores the result of applying FFT to data. proj_f array in the code.
+    filtered_freq_slice = (
+        non_slice_dims_shape[0] * (det_width // 2 + 1) * np.complex64().itemsize
     )
-    fftplan_size = freq_slice * 2
-    filtered_in_data = np.prod(non_slice_dims_shape) * np.float32().itemsize
 
-    # astra backprojection will generate an output array
-    astra_out_size = np.prod(output_dims) * np.float32().itemsize
+    batch = non_slice_dims_shape[0]
+    SLICES = 200  # dummy multiplier+divisor to pass large batch size threshold
+    fftplan_size = (
+        cufft_estimate_1d(
+            nx=det_width,
+            fft_type=CufftType.CUFFT_R2C,
+            batch=batch * SLICES,
+        )
+        / SLICES
+    )
+    ifftplan_size = (
+        cufft_estimate_1d(
+            nx=det_width,
+            fft_type=CufftType.CUFFT_C2R,
+            batch=batch * SLICES,
+        )
+        / SLICES
+    )
 
     tot_memory_bytes = int(
-        2 * in_slice_size
-        + filtered_in_data
-        + freq_slice
+        input_slice_size
+        + filtered_input_slice_size
+        + filtered_freq_slice
         + fftplan_size
```

Comment:
Testing with larger datasets (40gb as 80 doesn't work currently) I had to double the `fftplan_size` for it to pass. @yousefmoazzam do you know any reason what that should be the case or what is the current estimate missing? We know now that FFT filtering part is the most memory hungry one there, so we just looking at it. cheers

Reply:
I'll keep looking more closely, but on a first pass at the FBP code, something that sticks out to me is a question of if this line is being accounted for in the memory estimator https://github.com/dkazanc/ToMoBAR/blob/54137829b6326406e09f6ef9c95eb35c213838a7/tomobar/methodsDIR_CuPy.py#L135

The [cupy docs](https://docs.cupy.dev/en/stable/reference/generated/cupy.swapaxes.html) say that `swapaxes()` will return a view of the array. Non-trivial views of a multi-dimensional array; ie, anything that involves more than simply picking a subset with no fancy indexing will in general be rearranging the ordering of elements I think.

Swapping the order of two axes is a non-trivial operation as far as the arrangement of elements. So it'll probably rearrange elements in such a way that if the original element layout was C-contiguous, the layout of the elements in the view with the axes swapped won't be C-contiguous.

Forcing such a view of data to be C-contiguous feels like it'll most likely result in a copy, and in this case it's the entire data having its axes swapped and being forced to be C-contiguous.

Reply:
Thanks, this might explain why Jorg (I think) did initial estimate by [doubling](https://github.com/DiamondLightSource/httomo/blob/a5067f08965a44697aa31d593ee0686ed00cfbd4/httomo/methods_database/packages/external/httomolibgpu/supporting_funcs/recon/algorithm.py#L85) the size of the input slice.  Will give it a test...

Reply:
unfortunately even when doubling the `input_slice_size` there is not enough memory. And the OOM failure is inside  `filtersinc3D_cupy` part. [This](https://github.com/dkazanc/ToMoBAR/blob/54137829b6326406e09f6ef9c95eb35c213838a7/tomobar/fourier.py#L70) shouldn't create a copy?

Reply:
That's correct, that in-place multiplication shouldn't cause a copy of the array to be created.

It's tedious work, but at times when doing work on memory estimators before, I ended up using cupy's [`LineProfileHook`](https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.memory_hooks.LineProfileHook.html) to look at all the memory allocations being done that cupy is aware of.

It can be used like the hook used in the memory estimator tests, via `with`. That might be useful to do to see if there's any allocations happening in the cupy filtering that hasn't been accounted for.